### DISCUSSION TITLE
Exporting all types?

### DISCUSSION THREAD

**Discussion by rob-gordon** [joe-bell/cva#104] [Ideas]
I'm building some cool things around cva but one thing that makes it difficult is that only some of the types are exported from the library. Would there be any harm in exporting all of the internal types? Especially [Class Value](https://github.com/joe-bell/cva/blob/21652e5d42a0bef3e10eb1bd90bde9b48efe10ae/packages/class-variance-authority/src/types.ts#L2) and [Config](https://github.com/joe-bell/cva/blob/21652e5d42a0bef3e10eb1bd90bde9b48efe10ae/packages/class-variance-authority/src/index.ts#L40-L48). 
That would make it easier to extend cva or use it in different ways but maintain type-safety. I'm happy to open a PR if the maintainers are alright with it 👍🏻

**Comment by joe-bell** [comment]
Sorry, I don’t think this is something I want to export unless it’s something I feel comfortable maintaining 

Keeping it internal allows for more flexibility at the moment 

Happy to reconsider later

**Reply by rob-gordon** [reply]
I understand not wanting to add to the maintenance cost.

For context, type access allowed me to build a helper function for creating components directly from cva (built an [example](https://twitter.com/tone_row_/status/1618992027617865728) on a fork with types). I could imagine other people might find it useful too if it were a possibility.

**Reply by joe-bell** [reply]
Nice! [I intentionally chose not to include that kind of API](https://github.com/joe-bell/cva#why-dont-you-provide-a-styled-api), but glad to hear you’ve found something that works for you

I’ll let you know if I change my mind on exporting types! May reconsider for v1

**Reply by dolma** [reply]
Hi @joe-bell,

> Sorry, I don’t think this is something I want to export unless it’s something I feel comfortable maintaining
>
> Keeping it internal allows for more flexibility at the moment

Like @rob-gordon, I also understand not wanting to add to the maintenance cost but I respectfully disagree that there would be added maintenance cost in the first place. Here's why:

People who want to use the types will still use them regardless of what you do and are currently forced to either fork the repo just to add an `export` clause in front of the desired types (like @rob-gordon did), or have to manually duplicate the types into their code (like @ksandinunikum did). In which case you get the same risk as if the types were natively exported from the library: people's code can break if the types change internally and become incompatible with the ones they've duplicated in their code.

This means:
- people who don't need to use the types now, still wouldn't use them if they were exported either, and wouldn't be affected by any internal change anyway -> no change in flexibility
- people who need to use the types now are already finding ways to access them one way or another, and have to resort to approaches which are susceptible to breaking if internal changes are made anyway -> no change in flexibility
- people who forked the repo won't get affected either way -> still no change in flexibility

Essentially, in every scenario internal changes would have the exact same impact on people whether the types are exported or not, so no added flexibility whatsoever by not exporting them. However, not exporting them adds unnecessary pain for people who want to use the types.

**Comment by ksandinunikum** [comment]
I think this would be a great addition to CVA. I have also created typesafe extension utilities for CVA, and currently I have to redefine all the internal types. I think CVA is a really composable library fundamentally, but without also exposing the type definitions of its data structure, that composability will be limited to javascript only.
